src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/MemoryScope.java

```diff
@@ -181,6 +185,17 @@ protected Object clone() throws CloneNotSupportedException {
         throw new CloneNotSupportedException();
     }
 
+    /**
+     * Return a scoped allocator; the instance returned by this method is lazily created and then shared
+     * upon subsequent requests.
+     */
+    public SegmentAllocator allocator() {
```

Comment:
Potentially `MemoryScope` could implement `SegmentAllocator` then there is no need for this method?

And then on `SegmentAllocator`:
```
    static SegmentAllocator scoped(ResourceScope scope) {
        return (MemoryScope) Objects.requireNonNull(scope);
    }
```

Reply:
Yes - I guess I didn't want to blur the two concepts too much - if we go down that path I think users will be able to cast a ResourceScope into a SegmentAllocator?

Reply:
They could, but they should not since it's an implementation detail. (FWIW I don't think there is any impact to the integrity of the implementation.)